pkg/edgeview/src/tcp.go

```diff
@@ -389,6 +439,173 @@ func setAndStartProxyTCP(opt string) {
 	}
 }
 
+// Generate a random 32-byte symmetric key for kubeconfig file encryption
+func generateSymmetricKey() ([]byte, error) {
+	key := make([]byte, 32) // 256-bit key for AES-256
+	_, err := rand.Read(key)
+	if err != nil {
+		return nil, err
+	}
+	hexKey := hex.EncodeToString(key) // 64-character hexadecimal string
+
+	// Generate a random byte
+	var b [1]byte
+	_, err = rand.Read(b[:])
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the byte to an integer and take the modulus to get it within the range 0-31
+	start := int(b[0]) % 32
+
+	// Return a 32-character substring starting from the random index
+	return []byte(hexKey[start : start+32]), nil
+}
+
+func encryptSymmetricKey(symKey []byte, publicKeyFile string) ([]byte, error) {
+	pubKeyData, err := os.ReadFile(publicKeyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	parsed, _, _, _, err := ssh.ParseAuthorizedKey([]byte(pubKeyData))
+	if err != nil {
+		return nil, err
+	}
+	parsedCryptoKey := parsed.(ssh.CryptoPublicKey)
+	pubCrypto := parsedCryptoKey.CryptoPublicKey()
+	pub := pubCrypto.(*rsa.PublicKey)
+
+	encryptedBytes, err := rsa.EncryptPKCS1v15(
+		rand.Reader,
+		pub,
+		symKey,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return encryptedBytes, nil
+}
+
+func opensslEncryptConfig(symKey []byte, yamlString string) ([]byte, error) {
+	// save the config into a tmp file
+	idStr := strconv.Itoa(edgeviewInstID)
+	tmpCfgFile := "/tmp/ev-kubecfg-tmp." + idStr
+	tmpCfgFileEnc := "/tmp/ev-kubecfg-enc-tmp." + idStr
+	err := os.WriteFile(tmpCfgFile, []byte(yamlString), 0644)
+	if err != nil {
+		log.Errorf("opensslEncryptConfig: write tmp file error %v", err)
+		return nil, err
+	}
+
+	// openssl encrypt input kubeconfig file
+	name := "/usr/bin/openssl"
+	args := []string{"enc",
+		"-aes-256-cbc",
+		"-salt",
+		"-in",
+		tmpCfgFile,
+		"-out",
+		tmpCfgFileEnc,
+		"-k",
+		string(symKey),
+	}
+	cmd := exec.Command(name, args...)
```

Comment:
Why are we using `openssl` command line instead of the `crypto` pkg?

Reply:
I think i experimented some options then (more than half year ago), I could not seem to get other ways to work properly. this openssl works at least.

Reply:
I am a bit afraid that if we don't understand now why it does not work, then we will also later not understand if a user runs into a problem with this.

Reply:
Sure, but i didn't want to spend too much time on this, we can open a ticket to revisit the options on this

Reply:
I have done (too) much (painful) work with crypto. I am happy to take a stab at it.

Based on the command:

```bash
/usr/bin/openssl enc -aes-256-cbc -salt -k symKey -in - -out tmpCfgFileEnc
```

what you are trying to do is encrypt stdin using the cipher aes-256-cbc, adding a salt, using the key in `symKey` and saving it to `tmpCfgFileEnc`. Correct?

Reply:
How to perform openssl-compatible encryption/decryption with crypto can be found in https://github.com/Luzifer/go-openssl
We could also use that package, seems to be actively maintained.

Reply:
Oh, I didn't see that. I was ⅔ of the way there, but why add even a single file of our own custom, when there is OSS off the shelf available. By all means, do use that.

btw, did you see [the warning on it](https://github.com/Luzifer/go-openssl/blob/master/openssl.go#L165C11-L166)?

```go
// If you don't have a good reason to use this, please don't! For more information
// see this: https://en.wikipedia.org/wiki/Salt_(cryptography)#Common_mistakes
```

Reply:
What are we using this for? Is this something that we need to be using this vs something else? At the very least, pbkdf2 to generate the key?

Reply:
@milan-zededa good info, will look at this
@deitch basically, we need to copy the kubernetes kubeConfig to the laptop side for debuguser account, we don't want to show up this file as clear text file on user's laptop, so if the user has the SSH private key, they the user can read this and perform the remote 'kubectl' for the debugging session.
For this we first have a 32 bytes symmetric key, and we use the user's SSH public key info on the device, to encrypt this symmetric key, so this stdin feeding is the public key info. this encrypted key file and also the encrypted kubeconfig file both downloaded onto user's laptop.
User needs to use openssl and the SSH private key in order first to decrypt the symmetric key file, then the kubeconfig file. 
For the above, the user side must run 'openssl', and I was saying I was trying with the crypto in go and could not make it to work months ago.
If you have some tricks to do:
- on the remote system, use go crypto to encrypt the file with SSH pub key info
- copy the encrypted file to another system, and use SSH private key to decrypt that
-

Reply:
If I understood you correctly, the situation is as follows:

1. We want to encrypt some cleartext (doesn't matter what) on the device side, where the user can decrypt it on the laptop side
2. We want to use key material they already have, likely the key pair
3. We want to use tools they likely already have; they are unlikely to have any special eve tooling, and we do not want to provide it for them; likely they have openssl installed

So we create a shared symmetric key, use their already existing ssh public key to encrypt the symmetric key, use the symmetric key to encrypt the cleartext. The user then can use openssl (which they already have) with the ssh private key (which they already have) to decrypt the symmetric key, and use that to decrypt the encrypted text to cleartext.

If I understood correctly, I have 2 questions:

1. Why bother with the symmetric key? Using a public/private keypair to encrypt/decrypt (expensive) a randomly generated symmetric key (cheap) is useful when you have lots of material to encrypt/decrypt, e.g. TLS comms. If this is a one-time thing, why not just encrypt cleartext with the ssh public key?
2. Do we need to encrypt this at all? You wrote:

> we don't want to show up this file as clear text file on user's laptop

Why not? The vast majority of kubeconfig files are exactly that, in cleartext. Securing your kubeconfig files is generally considered beyond the scope of Kubernetes. Is this an actual problem we have, to make this _more_ secure than existing Kubernetes implementations?

Sure, there are more secured methods (think of how AWS might use credentials, although in the end, they are stored in your `~/.aws/credentials` file anyways, although that might plug into some form of secure on-system storage), but isn't that beyond scope?

Can we just say, "you generate it, the kubeconfig is stored at place "X", secure it however you normally secure sensitive material"?

Reply:
@deitch for 1) the file being encryption/description with the pub/priv key is limited to the size of the SSH key, and in this case, the kubeconfig exceeding the limit.
2) actually I'm not sure. If this forum agrees that we don't need to encrypt this info, I would happy to remove this, and it will be much simplified.

Reply:
> the file being encryption/description with the pub/priv key is limited to the size of the SSH key, and in this case, the kubeconfig exceeding the limit

OK. Makes sense.

> actually I'm not sure. If this forum agrees that we don't need to encrypt this info, I would happy to remove this, and it will be much simplified

Heh, sometimes the best engineering solution is, why bother doing it in the first place?  😆 

My vote is, unless we know we need it, and we know we need it now, I would steer clear. Bring us up to functional equivalence and interconnectedness with everything else, solve the "secure your kubeconfig" as a separate problem that may not even be ours. 🤷‍♂️

Reply:
I thought it would give us something like: EVE edge-node kubernetes cluster access is very secure comparing to the normal remote kubernetes access. But I think I can lower that goal too :-)